File: client/query.go
package client

import (
	"context"
	"iter"
	"net/http"
	"net/url"
)

// QueryOptions carry the FQL scope and pagination for a query.
type QueryOptions struct {
	ListOptions
	// Locale and Channel restrict scoped-attribute predicates to one scope;
	// zero matches base (unscoped) values.
	Locale  string
	Channel string
}

func (o QueryOptions) values(typ, fql string) url.Values {
	q := url.Values{}
	q.Set("type", typ)
	q.Set("q", fql)
	if o.Locale != "" {
		q.Set("locale", o.Locale)
	}
	if o.Channel != "" {
		q.Set("channel", o.Channel)
	}
	return q
}

// QueryPage runs one page of an FQL query over the given root type.
func (c *Client) QueryPage(ctx context.Context, typ, fql string, opts ...QueryOptions) (*Page[EntitySummary], error) {
	o := QueryOptions{}
	if len(opts) > 0 {
		o = opts[0]
	}
	return listPage[EntitySummary](ctx, c, "/query", o.values(typ, fql), o.ListOptions)
}

// Query runs an FQL query and iterates every matching entity across pages:
//
//	for row, err := range c.Query(ctx, "product", `price > 100`) {
//	    if err != nil { return err }
//	    ...
//	}
func (c *Client) Query(ctx context.Context, typ, fql string, opts ...QueryOptions) iter.Seq2[EntitySummary, error] {
	o := QueryOptions{}
	if len(opts) > 0 {
		o = opts[0]
	}
	return paginate(func(cursor string) (*Page[EntitySummary], error) {
		o.Cursor = cursor
		return c.QueryPage(ctx, typ, fql, o)
	})
}

// ValidateQuery parses and binds an FQL query without running it, returning a
// *APIError (code VALIDATION, with a "position" detail) when it is invalid.
func (c *Client) ValidateQuery(ctx context.Context, typ, fql string) error {
	return c.do(ctx, http.MethodPost, "/query/validate", nil, map[string]string{"type": typ, "q": fql}, nil)
}

File: client-ts/src/services/query.ts
import type { RequestOptions } from '../http.js'
import type { ListOptions, Page } from '../pagination.js'
import { paginate, paginatePages } from '../pagination.js'
import type { EntitySummary } from '../models.js'
import { pageOf, pageQuery, requestPart, Service } from './base.js'

/** The arguments of an FQL query. */
export interface QueryOptions extends ListOptions {
  /** Narrows a localizable attribute to one locale. */
  locale?: string | undefined
  /** Narrows a scopable attribute to one channel. */
  channel?: string | undefined
}

/** The result of validating a query without running it. */
export interface ValidateQueryResult {
  valid?: boolean
}

/**
 * FQL query operations.
 *
 * FQL filters entities of one root type: `price > 10 and status = "active"`.
 * A malformed expression answers 422, which this client raises as a VALIDATION
 * FlexitypeError with the position in `details`.
 */
export class QueryService extends Service {
  /** One page of the entities the expression matches. */
  async run(type: string, q: string, options: QueryOptions = {}): Promise<Page<EntitySummary>> {
    return pageOf(
      await this.http.request<Page<EntitySummary>>(
        'GET',
        '/query',
        {
          query: {
            type,
            q,
            ...pageQuery(options),
            locale: options.locale,
            channel: options.channel,
          },
        },
        requestPart(options),
      ),
    )
  }

  /** Every matching entity, following the cursor across pages. */
  runAll(type: string, q: string, options: QueryOptions = {}): AsyncGenerator<EntitySummary, void, undefined> {
    return paginate((cursor) => this.run(type, q, { ...options, cursor }))
  }

  /** Every matching page, for a caller that needs the page boundaries. */
  runPages(
    type: string,
    q: string,
    options: QueryOptions = {},
  ): AsyncGenerator<Page<EntitySummary>, void, undefined> {
    return paginatePages((cursor) => this.run(type, q, { ...options, cursor }))
  }

  /** Parses the expression without running it. */
  validate(type: string, q: string, options: RequestOptions = {}): Promise<ValidateQueryResult> {
    return this.http.request<ValidateQueryResult>('POST', '/query/validate', { body: { type, q } }, options)
  }
}
